File: Project2/Animal.cpp
#include "Animal.h"
#include "World.h"
#include <cstdlib>
#include <typeinfo>

Animal::Animal(const Vector2& pos, const int priority, const int strength, const int breedingCooldown, const char symbol)
	: Organism(pos, priority, strength, symbol), breedingCooldown(breedingCooldown)
{
}

void Animal::Action()
{
	Movement();

	currentBreedingCooldown--;
}

void Animal::Collide(Organism* other)
{
	world->Logs() << "Collision on " << pos << " (" << GetName() << ", " << other->GetName() << ")\n";
	if (GetName() == other->GetName())
	{
		Breed((Animal*)other);
	}
	else
	{
		other->Hit(this);
	}
}

void Animal::Movement()
{
	MoveTo(pos.GetNeighbor(rand() % 4));
}

void Animal::MoveTo(const Vector2& target)
{
	if (world->ContainsPos(target))
	{
		prevPos = pos;
		pos = target;
		world->Logs() << GetName() << " moved to " << pos << '\n';
	}
	else
	{
		world->Logs() << GetName() << " couldn't move to " << target << '\n';
	}
}

void Animal::Breed(Animal* other)
{
	GoBack();
	if (!world->HasEmptyNeighbor(other->GetPosition()))
	{
		world->Logs() << "Breeding failed on " << pos << " (not enough space)\n";
		return;
	}
	if (currentBreedingCooldown > 0 || other->currentBreedingCooldown > 0)
	{
		world->Logs() << "Breeding failed on " << pos << " (cooldown)\n";
		return;
	}

	Vector2 childPos = world->GetEmptyNeighbor(other->GetPosition());
	Animal* child = Clone(childPos);
	world->AddOrganism(child);

	currentBreedingCooldown = breedingCooldown;
	other->currentBreedingCooldown = other->breedingCooldown;
	child->currentBreedingCooldown = child->breedingCooldown;

	world->Logs() << "Breeding on " << child->pos << "\n";
}

File: Project2/WorldSimulator/Animal.java
package WorldSimulator;

import java.awt.*;
import java.util.ArrayList;
import java.util.Random;

public abstract class Animal extends Organism
{
    private final int breedingCooldown;
    private int currentBreedingCooldown;

    public Animal(Vector2 pos, int priority, int strength, int breedingCooldown, char symbol, Color color)
    {
        super(pos, priority, strength, symbol, color);
        this.breedingCooldown = breedingCooldown;
    }

    @Override
    public void Action()
    {
        Movement();

        currentBreedingCooldown--;
    }

    protected void Movement()
    {
        Random rand = new Random();
        ArrayList<Vector2> neighbors = world.GetNeighbors(pos);
        MoveTo(neighbors.get(rand.nextInt(neighbors.size())));
        //MoveTo(pos.GetNeighbor(rand.nextInt(4)));
    }

    protected void MoveTo(Vector2 target)
    {
        if (world.ContainsPos(target))
        {
            prevPos = pos;
            pos = target;
            world.AddLog(this + " moved to " + pos + '\n');
        }
        else
        {
            world.AddLog(this + " couldn't move to " + target + '\n');
        }
    }

    @Override
    public void Collide(Organism organism)
    {
        world.AddLog("Collision on " + pos + " (" + this + ", " + organism + ")\n");
        if (toString().equals(organism.toString()))
        {
            Breed((Animal)organism);
        }
        else
        {
            organism.Hit(this);
        }
    }

    private void Breed(Animal other)
    {
        GoBack();
        if (!world.HasEmptyNeighbor(other.getPos()))
        {
            world.AddLog("Breeding failed on " + pos + " (not enough space)\n");
            return;
        }
        if (currentBreedingCooldown > 0 || other.currentBreedingCooldown > 0)
        {
            world.AddLog("Breeding failed on " + pos + " (cooldown)\n");
            return;
        }

        Vector2 childPos = world.GetRandomEmptyNeighbor(other.getPos());
        Animal child = (Animal)clone();
        child.pos = childPos;
        world.AddOrganism(child);

        currentBreedingCooldown = breedingCooldown;
        other.currentBreedingCooldown = other.breedingCooldown;
        child.currentBreedingCooldown = child.breedingCooldown;

        world.AddLog( "Breeding on " + child.pos + "\n");
    }
}
